[BUG] ESM: Cannot find module '@playwright/test/types/...' or its corresponding type declarations.

**Context**
- Playwright Version: @playwright/test@1.24.2
- Operating System: Mac with M1 chip
- Node.js version: v16.16.0
- Browser: n/a
- Extra: TypeScript 4.7.4 with `"type": "module"` and `"module": "NodeNext"`

System:
 - OS: macOS 12.5
 - Memory: 1.06 GB / 32.00 GB

Binaries:
 - Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
 - Yarn: 1.22.19 - ~/.yarn/bin/yarn
 - npm: 8.11.0 - ~/.nvm/versions/node/v16.16.0/bin/npm

Languages:
 - Bash: 3.2.57 - /bin/bash

**Code Snippet**

```ts
import { test, expect } from '@playwright/test';
```

**Describe the bug**

With `"type": "module"` and `"module": "NodeNext"` or `"module": "node16"` set, TypeScript throws two `TS2307` errors on `@playwright/test`:

```bash
➜ npm run tsc

> pw-esm-repro@0.0.0-development tsc
> tsc

node_modules/@playwright/test/types/test.d.ts:3027:34 - error TS2307: Cannot find module '@playwright/test/types/expect-types' or its corresponding type declarations.

3027 import type * as expectType from '@playwright/test/types/expect-types';
                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@playwright/test/types/test.d.ts:3028:28 - error TS2307: Cannot find module '@playwright/test/types/testReporter' or its corresponding type declarations.

3028 import type { Suite } from '@playwright/test/types/testReporter';
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors in the same file, starting at: node_modules/@playwright/test/types/test.d.ts:3027
```

This seems possibly related to https://github.com/microsoft/playwright/issues/14971?

**Repro**

1. Clone https://github.com/jgerigmeyer/pw-esm-repro
2. Checkout `ts-npm` branch
3. `npm i`
4. `npm run tsc`

--- a/packages/playwright-core/types/structs.d.ts
+++ b/packages/playwright-core/types/structs.d.ts
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-import { JSHandle, ElementHandle, Frame, Page, BrowserContext } from 'playwright-core';
+import { JSHandle, ElementHandle, Frame, Page, BrowserContext } from './types';
 
 /**
  * Can be converted to JSON

--- a/packages/playwright-core/types/types.d.ts
+++ b/packages/playwright-core/types/types.d.ts
@@ -14,12 +14,12 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Protocol } from 'playwright-core/types/protocol';
 import { ChildProcess } from 'child_process';
 import { EventEmitter } from 'events';
 import { Readable } from 'stream';
 import { ReadStream } from 'fs';
-import { Serializable, EvaluationArgument, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from 'playwright-core/types/structs';
+import { Protocol } from './protocol';
+import { Serializable, EvaluationArgument, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from './structs';
 
 type PageWaitForSelectorOptionsNotHidden = PageWaitForSelectorOptions & {
   state?: 'visible'|'attached';

--- a/packages/playwright-test/types/test.d.ts
+++ b/packages/playwright-test/types/test.d.ts
@@ -3042,8 +3042,8 @@ export interface PlaywrightTestArgs {
 export type PlaywrightTestProject<TestArgs = {}, WorkerArgs = {}> = Project<PlaywrightTestOptions & TestArgs, PlaywrightWorkerOptions & WorkerArgs>;
 export type PlaywrightTestConfig<TestArgs = {}, WorkerArgs = {}> = Config<PlaywrightTestOptions & TestArgs, PlaywrightWorkerOptions & WorkerArgs>;
 
-import type * as expectType from '@playwright/test/types/expect-types';
-import type { Suite } from '@playwright/test/types/testReporter';
+import type * as expectType from './expect-types';
+import type { Suite } from './testReporter';
 
 type AsymmetricMatcher = Record<string, any>;
 

--- a/packages/playwright-test/types/testReporter.d.ts
+++ b/packages/playwright-test/types/testReporter.d.ts
@@ -15,8 +15,8 @@
  * limitations under the License.
  */
 
-import type { FullConfig, FullProject, TestStatus, TestError, Metadata } from '@playwright/test';
-export type { FullConfig, TestStatus, TestError } from '@playwright/test';
+import type { FullConfig, FullProject, TestStatus, TestError, Metadata } from './test';
+export type { FullConfig, TestStatus, TestError } from './test';
 
 /**
  * `Suite` is a group of tests. All tests in Playwright Test form the following hierarchy:

--- a/tests/installation/typescript-types.spec.ts
+++ b/tests/installation/typescript-types.spec.ts
@@ -35,3 +35,26 @@ test('typescript types should work', async ({ exec, tsc, writeFiles }) => {
 
   await tsc('playwright-test-types.ts');
 });
+
+test('typescript types should work with module: NodeNext', async ({ exec, tsc, writeFiles }) => {
+  // module: NodeNext got added in TypeScript 4.7
+  await exec('npm i --foreground-scripts typescript@4.7 @types/node@18');
+  const libraryPackages = [
+    'playwright',
+    'playwright-core',
+    'playwright-firefox',
+    'playwright-webkit',
+    'playwright-chromium',
+  ];
+  await exec('npm i @playwright/test', ...libraryPackages, { env: { PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD: '1' } });
+
+  for (const libraryPackage of libraryPackages) {
+    const filename = libraryPackage + '.ts';
+    await writeFiles({
+      [filename]: `import { Page } from '${libraryPackage}';`,
+    });
+    await exec('npx', '-p', 'typescript@4.7', 'tsc', '--module nodenext', filename);
+  }
+
+  await exec('npx', '-p', 'typescript@4.7', 'tsc', '--module nodenext', 'playwright-test-types.ts');
+});

--- a/utils/generate_types/overrides-test.d.ts
+++ b/utils/generate_types/overrides-test.d.ts
@@ -252,8 +252,8 @@ export interface PlaywrightTestArgs {
 export type PlaywrightTestProject<TestArgs = {}, WorkerArgs = {}> = Project<PlaywrightTestOptions & TestArgs, PlaywrightWorkerOptions & WorkerArgs>;
 export type PlaywrightTestConfig<TestArgs = {}, WorkerArgs = {}> = Config<PlaywrightTestOptions & TestArgs, PlaywrightWorkerOptions & WorkerArgs>;
 
-import type * as expectType from '@playwright/test/types/expect-types';
-import type { Suite } from '@playwright/test/types/testReporter';
+import type * as expectType from './expect-types';
+import type { Suite } from './testReporter';
 
 type AsymmetricMatcher = Record<string, any>;
 

--- a/utils/generate_types/overrides-testReporter.d.ts
+++ b/utils/generate_types/overrides-testReporter.d.ts
@@ -14,8 +14,8 @@
  * limitations under the License.
  */
 
-import type { FullConfig, FullProject, TestStatus, TestError, Metadata } from '@playwright/test';
-export type { FullConfig, TestStatus, TestError } from '@playwright/test';
+import type { FullConfig, FullProject, TestStatus, TestError, Metadata } from './test';
+export type { FullConfig, TestStatus, TestError } from './test';
 
 export interface Suite {
   project(): FullProject | undefined;

--- a/utils/generate_types/overrides.d.ts
+++ b/utils/generate_types/overrides.d.ts
@@ -13,12 +13,12 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Protocol } from 'playwright-core/types/protocol';
 import { ChildProcess } from 'child_process';
 import { EventEmitter } from 'events';
 import { Readable } from 'stream';
 import { ReadStream } from 'fs';
-import { Serializable, EvaluationArgument, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from 'playwright-core/types/structs';
+import { Protocol } from './protocol';
+import { Serializable, EvaluationArgument, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from './structs';
 
 type PageWaitForSelectorOptionsNotHidden = PageWaitForSelectorOptions & {
   state?: 'visible'|'attached';

